File: InClass/refactored Java demo/src/Card.java
/**
 * models a playing card
 */
public class Card {

    // constants use the keyword 'final'
    private final String[] SUITS = {
        "Hearts", "Clubs", "Spades", "Diamonds"
    };
    private final int DEFAULT_RANK = 14;
    private final String DEFAULT_SUIT = "Hearts";

    private final int ACE = 14;
    private final int KING = 13;
    private final int QUEEN = 12;
    private final int JACK = 11;

    private final int HEARTS = 0;
    private final int CLUBS = 1;
    private final int SPADES = 2;
    private final int DIAMONDS = 3;

    private final int RANK = 0;
    private final int SUIT = 1;

    // DECLARE instance variables

//    private int rank;    // 2-14
//    private String suit; // "Clubs", etc.
    private int[] card_info;

    /**
     * non default-constructor == constructor with params
     *
     * @param newRank int in range 2-14
     * @param newSuit String spelled out like "Clubs"
     */
    public Card(int newRank, String newSuit) {
        card_info = new int[2];
        card_info[RANK] = newRank;
        card_info[SUIT] = stringSuitToInt(newSuit);
    }

    /**
     * default constructor == constructor with no parameters
     */
    public Card() {
        card_info = new int[2];
        card_info[RANK] = DEFAULT_RANK;
        card_info[SUIT] = stringSuitToInt(DEFAULT_SUIT);
    }

    /**
     * non-default constructor
     *
     * @param newSuit suit spelled out like "Clubs"
     */
    public Card(String newSuit) {
        card_info = new int[2];
        card_info[RANK] = DEFAULT_RANK;
        card_info[SUIT] = stringSuitToInt(newSuit);

//        this.rank = DEFAULT_RANK;
//        this.suit = newSuit;
    }

    /**
     * getter for the rank
     *
     * @return int from 2-14
     */
    public int getRank() {
        return card_info[RANK];
    }

    /**
     * getter for suit
     *
     * @return String denoting full suit like "Spades"
     */
    public String getSuit() {
        return SUITS[card_info[SUIT]];
//        return suit;
    }

    /**
     * return this Card as a printable string
     *
     * @return card's full name like "Ace of Hearts"
     */
    public String toString() {
        int rank = this.getRank();
        String rankToPrint;
        if (rank == ACE) {
            rankToPrint = "Ace";
        }
        else if (rank == KING) {
            rankToPrint = "King";
        }
        else if (rank == QUEEN) {
            rankToPrint = "Queen";
        }
        else if (rank == JACK) {
            rankToPrint = "Jack";
        }
        else {
            rankToPrint = "" + rank;
        }
        return rankToPrint + " of " + getSuit();
    }

    private int stringSuitToInt(String suit) {
        if (suit == "HEARTS"){
            return HEARTS;
        }else if (suit == "SPADES"){
            return SPADES;
        }else if (suit == "DIAMONDS"){
            return DIAMONDS;
        } else if (suit == "CLUBS") {
            return CLUBS;
        }else{
            return -1;
        }
    }
}

File: InClass/robustness demo/card.py
"""models a playing card"""

from constants import *

class Card:
    def __init__(self, rank=1, suit="Spades"):
        """
        constructor; defaults to Ace of Spades if no arguments given
        :param rank: int from 1-13 (Ace = 1)
        :param suit: string of suit completely spelled out
                     with initial letter capitalized (like "Spades")
        """
        if MIN_RANK <= rank <= MAX_RANK and suit in SUITS:
            self.__rank = rank
            self.__suit = suit
        else:
            self.__init__()

    def get_rank(self):
        """return MY rank as an int 1-13"""
        return self.__rank


    def get_suit(self):
        """return MY suit (like 'Spades')"""
        return self.__suit


    def __str__(self):
        """return MYSELF as a printable string"""
        rank = self.get_rank()
        if rank == 1:
            rank_string = "Ace"
        elif rank == 11:
            rank_string = "Jack"
        elif rank == 12:
            rank_string = "Queen"
        elif rank == 13:
            rank_string = "King"
        else:
            rank_string = str(rank)
        return rank_string + " of " + self.get_suit()


if __name__ == "__main__":
    card = Card()
    print(card)
    card2 = Card(3,"Hearts")
    print(card2)
    card3 = Card(0,"Clubs")
    print(card3)
    card4 = Card(2,"Ham")
    print(card4)